connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufUtils.scala

```diff
@@ -196,30 +194,52 @@ private[sql] object ProtobufUtils extends Logging {
       fileDescriptorSet = DescriptorProtos.FileDescriptorSet.parseFrom(dscFile)
     } catch {
       case ex: InvalidProtocolBufferException =>
-        // TODO move all the exceptions to core/src/main/resources/error/error-classes.json
-        throw new RuntimeException("Error parsing descriptor byte[] into Descriptor object", ex)
+        throw QueryCompilationErrors.descrioptorParseError(ex)
       case ex: IOException =>
-        throw new RuntimeException(
-          "Error reading Protobuf descriptor file at path: " +
-            descFilePath,
-          ex)
+        throw QueryCompilationErrors.cannotFindDescriptorFileError(descFilePath, ex)
     }
-
-    val descriptorProto: DescriptorProtos.FileDescriptorProto = fileDescriptorSet.getFile(0)
     try {
-      val fileDescriptor: Descriptors.FileDescriptor = Descriptors.FileDescriptor.buildFrom(
-        descriptorProto,
-        new Array[Descriptors.FileDescriptor](0))
+      val fileDescriptorProtoIndex = createDescriptorProtoMap(fileDescriptorSet)
+      val fileDescriptor: Descriptors.FileDescriptor =
+        buildFileDescriptor(fileDescriptorSet.getFileList.asScala.last, fileDescriptorProtoIndex)
       if (fileDescriptor.getMessageTypes().isEmpty()) {
-        throw new RuntimeException("No MessageTypes returned, " + fileDescriptor.getName());
+        throw QueryCompilationErrors.noProtobufMessageTypeReturnError(fileDescriptor.getName())
       }
       fileDescriptor
     } catch {
       case e: Descriptors.DescriptorValidationException =>
-        throw new RuntimeException("Error constructing FileDescriptor", e)
+        throw QueryCompilationErrors.failedParsingDescriptorError(e)
     }
   }
 
+  /**
+   * Recursively constructs file descriptors for all dependencies for given
+   * FileDescriptorProto and return.
+   */
+  private def buildFileDescriptor(
+    fileDescriptorProto: FileDescriptorProto,
+    fileDescriptorProtoMap: Map[String, FileDescriptorProto]): Descriptors.FileDescriptor = {
+    val fileDescriptorList = fileDescriptorProto.getDependencyList().asScala.map { dependency =>
+      fileDescriptorProtoMap.get(dependency) match {
+        case Some(dependencyProto) =>
+          buildFileDescriptor(dependencyProto, fileDescriptorProtoMap)
+        case None =>
+          throw QueryCompilationErrors.protobufDescriptorDependencyError(dependency)
+      }
+    }
+    Descriptors.FileDescriptor.buildFrom(fileDescriptorProto, fileDescriptorList.toArray)
+  }
+
+  /**
+   * Returns a map from descriptor proto name as found inside the descriptors to protos.
+   */
+  private def createDescriptorProtoMap(
+    fileDescriptorSet: FileDescriptorSet): Map[String, FileDescriptorProto] = {
+    fileDescriptorSet.getFileList().asScala.map { descriptorProto =>
+      descriptorProto.getName() -> descriptorProto
```

Comment:
Btw, What happens if there are two different files with the same name? E.g. 
`protoc ... a/b/events.proto x/y/events.proto`

Reply:
@rangadi protoc command would throw the below error, protoc won't allow duplicate imports. 

```
protoc --java_out=./ functions_suite.proto
basicmessage_1.proto:30:9: "org.apache.spark.sql.protobuf.protos.BasicMessage.id" is already defined in file "basicmessage.proto".
basicmessage_1.proto:31:10: "org.apache.spark.sql.protobuf.protos.BasicMessage.string_value" is already defined in file "basicmessage.proto".
basicmessage_1.proto:32:9: "org.apache.spark.sql.protobuf.protos.BasicMessage.int32_value" is already defined in file "basicmessage.proto".
basicmessage_1.proto:33:9: "org.apache.spark.sql.protobuf.protos.BasicMessage.int64_value" is already defined in file "basicmessage.proto".
basicmessage_1.proto:34:10: "org.apache.spark.sql.protobuf.protos.BasicMessage.double_value" is already defined in file "basicmessage.proto".
```

Reply:
The command is using only single proto file. Not the same as the example above, right?

Reply:
@rangadi yeah correct but functions_suite.proto is including both basicmessage.proto and basicmessage_1.proto, wouldn't it equal the same as above? 
```
import "timestamp.proto";
import "duration.proto";
import "basicmessage.proto";
import "basicmessage_1.proto";
```

Reply:
That is not what I am referring to. This is not really a problem since it won't even compile. 

Notice two files as input for `protoc` above. May be that will since `name` will include the path.

Reply:
@rangadi Yes, protoc allowed two .proto files from different paths, but the last .proto file overrides the first one if there are any conflicts. but It is entirely up to the user to determine how to create a .desc file right?